asserts/ifacedecls.go

```diff
@@ -501,6 +504,88 @@ func compileDeviceScopeConstraint(cMap map[string]interface{}, context string) (
 	}, nil
 }
 
+type nameMatcher interface {
+	match(name string, special map[string]string) error
+}
+
+var (
+	validSpecialNameConstraint = regexp.MustCompile("^\\$[A-Z][A-Z0-9_]*$")
+)
+
+func compileNameMatcher(whichName string, v interface{}) (nameMatcher, error) {
+	s, ok := v.(string)
+	if !ok {
+		return nil, fmt.Errorf("%s constraint entry must be a regexp or special $ value", whichName)
+	}
+	if strings.HasPrefix(s, "$") {
+		if !validSpecialNameConstraint.MatchString(s) {
+			return nil, fmt.Errorf("%s constraint entry special value %q is invalid", whichName, s)
+		}
+		return specialNameMatcher{special: s}, nil
+	}
+	rx, err := regexp.Compile("^(" + s + ")$")
```

Comment:
Do we allow a full Go regex or should it be more conservative and use `CompilePOSIX()`?

Should we complain about `\s`, `' '`, `\\`?

Reply:
we have allowed full Go regexes so far in the same context see other code here.  We haven't done a lot of extra checks so far. Our worried about people matching for spaces given that this must be names?

Reply:
For context this is similar preexisting code:

https://github.com/snapcore/snapd/blob/master/asserts/ifacedecls.go#L240

Reply:
Yes, like accidentally creating a regexp that can't really match a valid slot or plug name. Although, I see now there's similar code for attribute constraints, so probably not worth the extra complexity.

Reply:
I could look for explicit whitespace, more than that seems overkill complicated, there is a lot of ways to write a regexp (for better or worse).

Reply:
Yeah, probably not worth to dwell on it.

Reply:
@bboozzoo I did this https://github.com/snapcore/snapd/pull/7812/commits/1ceddf926d8062ab9092b706d09bf8cab2f13d0f